### PR TITLE
convert_sync_batch_norm to SyncBatchNorm

### LINKED ISSUE #18382: [FR] make torch.nn.utils.convert_sync_batchnorm a classmethod of SyncBatchNorm
The function really goes together with SyncBatchNorm, so I think it makes more sense for them to live together. Users would have easier time finding/discovering it too.

https://github.com/pytorch/pytorch/blob/39669316a6fc31c26f5ef1be5c0a0fe862d661cf/torch/nn/utils/sync_batch_norm.py#L4

cc @jjsjann123

### PR DESCRIPTION
**Pull request by iArunava**
Closes #18382 

Please let me know if any changes are required.

### CODE REVIEW

**Review by ssnl — CHANGES_REQUESTED** [review]
There are other call sites of `convert_sync_batchnorm` that you should fix https://github.com/pytorch/pytorch/search?q=convert_sync_batchnorm&unscoped_q=convert_sync_batchnorm

File: torch/nn/modules/_functions.py [hunk 1]
```diff
@@ -96,3 +96,47 @@ def backward(self, grad_output):
             grad_bias = None
 
         return grad_input, grad_weight, grad_bias, None, None, None, None, None, None
+
+    @staticmethod
```
Comment (SsnL) [inline torch/nn/modules/_functions.py]: This should be a class method. Line 128 should use `cls` instead

File: torch/nn/modules/_functions.py [hunk 2]
```diff
@@ -96,3 +96,47 @@ def backward(self, grad_output):
...
+                                                   module.track_running_stats,
+                                                   process_group)
+            if module.affine:
+                module_output.weight.data = module.weight.data.clone().detach()
+                module_output.bias.data = module.bias.data.clone().detach()
+            module_output.running_mean = module.running_mean
+            module_output.running_var = module.running_var
+            module_output.num_batches_tracked = module.num_batches_tracked
+        for name, child in module.named_children():
+            module_output.add_module(name, convert_sync_batchnorm(child))
```
Comment (SsnL) [inline torch/nn/modules/_functions.py]: This recursive call needs to be fixed.

**Review by ssnl — COMMENTED** [review]
next time you submit/update a PR for review. please make sure that the relevant tests pass locally first.

File: torch/nn/modules/_functions.py (hunk 2 above)
Comment (iArunava) [reply torch/nn/modules/_functions.py]: True. Totally missed that. Also, to make sure that the relevant tests pass, I just need to use the relevant test files in the `tests` directory right?

File: torch/nn/modules/_functions.py (hunk 1 above)
Comment (iArunava) [reply torch/nn/modules/_functions.py]: Will Fix that.

File: torch/nn/modules/_functions.py (hunk 2 above)
Comment (fmassa) [reply torch/nn/modules/_functions.py]: @iArunava yes, just run `python test_torch.py` etc and it should be fine

Comment (iArunava) [reply torch/nn/modules/_functions.py]: I have made the changes required, also all the places here: https://github.com/pytorch/pytorch/search?q=convert_sync_batchnorm&unscoped_q=convert_sync_batchnorm 

`torch.nn.convert_sync_batch_norm` is replaced by `torch.nn.SyncBatchNorm.convert_sync_batch_norm` as its a `@classmethod` Is should be accessible in that way. Is that okay?

And before pushing I want to run the tests, but I am not able to run the tests, I keep getting cannot import module `istuple` from the `torch._six.py` file, I can't find the reason why, everything else from `_six.py` is imported okay. Can you please help @fmassa ? I am quite sure I am missing something here.

**Comment by iArunava** [comment]
I pushed without testing sorry for that. But I was unable to get the tests working. I expected it will work. I tried uninstalling torch and trying to get into develop mode as said in `CONTRIBUTING.md` but the build kept failing. Can you please help me as how to get the `develop` mode ? Thanks! I will fix this ASAP.

**Comment by SsnL** [comment]
It's okay. For a small change like this, you can use the CI to fix things too. I'll tag this as WIP. When you are done and this is ready for review. Feel free to remove "WIP" and/or tag us :)

For the long term and bigger changes, it would still be helpful to figure out local installation.

File: torch/nn/modules/_functions.py [hunk 3]
```diff
@@ -96,3 +96,47 @@ def backward(self, grad_output):
             grad_bias = None
 
         return grad_input, grad_weight, grad_bias, None, None, None, None, None, None
+
+    @classmethod
+    def convert_sync_batchnorm(cls, module, process_group=None):
```
Comment (SsnL) [inline torch/nn/modules/_functions.py]: hey you should add this on the module rather than the autograd function!

Comment (iArunava) [reply torch/nn/modules/_functions.py]: Hey if I get you correctly, then you are asking to move the `convert_sync_batchnorm` function to in the `SyncBatchNorm` function here: https://github.com/pytorch/pytorch/blob/0829ef00dd5b62f7850290c722e9eaa49d45b491/torch/nn/modules/batchnorm.py#L322 Please correct me if I am wrong.

**Comment by iArunava** [comment]
@SsnL can you please check once, I have finally built pytorch on my machine. And I have changed the `convert_sync_batch_norm` from the autograd to the module and have checked it and it executed successfully. `nn.SyncBatchNorm.convert_sync_batch_norm(module)` works.
 Also, I ran the `python test/run_test.py` it shows everything is okay.

**Comment by iArunava** [comment]
@SsnL Please check! :)

**Review by soumith — APPROVED** [review]
looks good now. thank you!
